### ISSUE TITLE
How to specify peer IP correctly in Docker container for NCCL backend(PyTorch distributed)? 

### ISSUE THREAD

**Issue by gemfield** [NVIDIA/nccl#331]
Suppose I have two nodes A and B in a subnet, and I launched one Docker container on each node(thus CA and CB) with below command:
```bash
docker run -it -p 27030:27030 ...
```
As you can see, I didn't use "--network host" parameter for purpose.
Now I have below IPs:
```
A: 192.168.0.91
B: 192.168.0.34
CA: 172.17.0.3
CB: 172.17.0.5
```
After I enabled below ENVs on CA and CB:
```bash
export NCCL_SOCKET_IFNAME=eth0 #172.17.0.3 on CA
export NCCL_IB_DISABLE=1
export NCCL_DEBUG=INFO
export NCCL_DEBUG_SUBSYS=ALL
```
In the trainer script, I have
```bash
dist.init_process_group(backend='nccl', init_method='tcp://192.168.0.91:27030',
                                world_size=args.world_size, rank=args.rank)
```
As CA is a master, on CB, the trainer script has below outputs:
```bash
71d46aa8dab5:5241:5262 [2] NCCL INFO Call to connect returned Connection refused, retrying
71d46aa8dab5:5241:5262 [2] NCCL INFO Call to connect returned Connection refused, retrying
71d46aa8dab5:5241:5262 [2] NCCL INFO Call to connect returned Connection refused, retrying

71d46aa8dab5:5241:5262 [2] include/socket.h:390 NCCL WARN Connect to 172.17.0.3<37551> failed : Connection refused
71d46aa8dab5:5241:5262 [2] NCCL INFO bootstrap.cc:100 -> 2
71d46aa8dab5:5241:5262 [2] NCCL INFO bootstrap.cc:326 -> 2
71d46aa8dab5:5241:5262 [2] NCCL INFO init.cc:695 -> 2
71d46aa8dab5:5241:5262 [2] NCCL INFO init.cc:951 -> 2
71d46aa8dab5:5241:5262 [2] NCCL INFO misc/group.cc:69 -> 2 [Async thread]
Traceback (most recent call last):
  File "main.py", line 425, in <module>
    main()
  File "main.py", line 109, in main
    mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, args))
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 119, in join
    raise Exception(msg)
Exception: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap
    fn(i, *args)
  File "/home/gemfield/pytorch-examples/imagenet/main.py", line 151, in main_worker
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 285, in __init__
    self.broadcast_bucket_size)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 483, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1587428398394/work/torch/lib/c10d/ProcessGroupNCCL.cpp:514, unhandled system error, NCCL version 2.4.8
```

My questions are:
1, How to make trainer script on CB get the correct IP:PORT, i.e. 192.168.0.91 rather than 172.17.0.3 ?
2, Is there  an elegant way to make CA communicate with CB with pure docker bridge network(i.e. not use --network=host)?

**Comment by kwen2501** [comment]
Hi, what is the IP address of eth0 in the containers? If it is 172.17...., then NCCL would be instructed to use that IP address as you've set `NCCL_SOCKET_IFNAME=eth0`

**Comment by gemfield** [comment]
@kwen2501 Yes, it is 172.17.0.3.

**Comment by sjeaugey** [comment]
The problem seems that the two containers cannot communicate with each other (CA cannot talk to CB). NCCL needs the two containers to be able to open a socket on a port and connect to it on the other side. Can you open a socket on container A (`nc -l 35571`) and connect to it from container B (`nc 172.17.0.3 35571`) ?

**Comment by gemfield** [comment]
@sjeaugey What I'm using is a typical docker bridge network, the commucation should behave like this:
CB <-----> A <------> CA
i.e. CB should communicate with CA by 192.168.0.91:27030 rather than 172.17.0.3:27030 (like DNAT). There has no direct network access between CA and CB. 
I have binded port 27030 on CA(container) to A (Host), so I gave ```init_method='tcp://192.168.0.91:27030'``` in trainer script on CB.

**Comment by kwen2501** [comment]
I am not sure whether PyTorch passes`192.168.0.91` address down to NCCL and how. Perhaps that would be a better question for the PyTorch community. Regularly, NCCL bootstraps based on the socket interface (in this case, eth0), that's why the 172.17 addresses were used in the log.

That said, if you want to experiment with your docker bridge network and use 192.168 addresses, you may try setting `NCCL_COMM_ID=192.168.0.91:27030` and **not** set `NCCL_SOCKET_IFNAME`. I am not sure if you would need to set `NCCL_COMM_ID` in both containers, that depends on whether the 192.168 address is visible both containers. 

In case the 192.168 address is not visible in container A, there would still be a problem I think, because during a second handshake the processes in container A would still provide a connection handle based on the 172.17 address, which is not visible to container B.

**Comment by sjeaugey** [comment]
Even if the bootstrap works, the communication wouldn't. NCCL needs both containers to be able to communicate directly to each other, not just have one port on the host which both containers can access.

Now, I'm no docker expert but I thought that was what a docker bridge would do : give an IP address to each container (172.16.a.b) and bridge network packets between the containers. @gemfield can you point to documentation on what you are trying to achieve and what a "typical docker bridge network" is ? I could only find this page : https://docs.docker.com/network/bridge/ which states : "in terms of Docker, a bridge network uses a software bridge **which allows containers connected to the same bridge network to communicate**".
So as I understand it, there should be no need to open a port or even talk to the host, CA and CB should be able to communicate directly since the bridge should relay all network traffic between the containers.

**Comment by gemfield** [comment]
@kwen2501 With below ENVs set on CA and CB:
```bash
#CA 
NCCL_IB_DISABLE=1
NCCL_VERSION=2.4.8
NCCL_COMM_ID=192.168.0.91:27030
NCCL_DEBUG_SUBSYS=ALL
NCCL_DEBUG=INFO

#CB 
NCCL_IB_DISABLE=1
NCCL_VERSION=2.4.8
NCCL_COMM_ID=192.168.0.34:27030
NCCL_DEBUG_SUBSYS=ALL
NCCL_DEBUG=INFO
```

I got following errors on CA:
```bash
e82c50601d13:213:213 [0] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>
e82c50601d13:213:213 [0] NCCL INFO Bootstrap : Using [0]eth0:172.17.0.3<0>
e82c50601d13:213:213 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
e82c50601d13:213:213 [0] NCCL INFO NCCL_IB_DISABLE set by environment to 1.

e82c50601d13:213:213 [0] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>
e82c50601d13:213:213 [0] NCCL INFO NET/Socket : Using [0]eth0:172.17.0.3<0>

e82c50601d13:213:213 [0] include/socket.h:333 NCCL WARN Call to bind failed : Cannot assign requested address
e82c50601d13:213:213 [0] NCCL INFO bootstrap.cc:91 -> 2
e82c50601d13:213:213 [0] NCCL INFO bootstrap.cc:245 -> 2
e82c50601d13:213:213 [0] NCCL INFO init.cc:970 -> 2

e82c50601d13:214:214 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>
e82c50601d13:214:214 [1] NCCL INFO Bootstrap : Using [0]eth0:172.17.0.3<0>
e82c50601d13:214:214 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
e82c50601d13:214:214 [1] NCCL INFO NCCL_IB_DISABLE set by environment to 1.

e82c50601d13:214:214 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>
e82c50601d13:214:214 [1] NCCL INFO NET/Socket : Using [0]eth0:172.17.0.3<0>
e82c50601d13:214:228 [1] NCCL INFO Setting affinity for GPU 1 to 3f

e82c50601d13:214:228 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>

e82c50601d13:214:228 [1] bootstrap.cc:83 NCCL WARN NET/Socket : No usable listening interface found
e82c50601d13:214:228 [1] NCCL INFO bootstrap.cc:312 -> 2
e82c50601d13:214:228 [1] NCCL INFO init.cc:695 -> 2
e82c50601d13:214:228 [1] NCCL INFO init.cc:951 -> 2
e82c50601d13:214:228 [1] NCCL INFO misc/group.cc:69 -> 2 [Async thread]
Traceback (most recent call last):
  File "main.py", line 425, in <module>
    main()
  File "main.py", line 109, in main
    mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, args))
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 119, in join
    raise Exception(msg)
Exception: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap
    fn(i, *args)
  File "/home/gemfield/github/pytorch-examples/imagenet/main.py", line 151, in main_worker
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 285, in __init__
    self.broadcast_bucket_size)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 483, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1587428398394/work/torch/lib/c10d/../c10d/NCCLUtils.hpp:82, unhandled system error, NCCL version 2.4.8
```

AND following errors on CB:
```bash
71d46aa8dab5:6002:6002 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.34<27030>
71d46aa8dab5:6001:6001 [0] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.34<27030>
71d46aa8dab5:6001:6001 [0] NCCL INFO Bootstrap : Using [0]eth0:172.17.0.5<0>
71d46aa8dab5:6002:6002 [1] NCCL INFO Bootstrap : Using [0]eth0:172.17.0.5<0>
71d46aa8dab5:6001:6001 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
71d46aa8dab5:6002:6002 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).

71d46aa8dab5:6001:6001 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]

71d46aa8dab5:6002:6002 [1] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]


71d46aa8dab5:6002:6002 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.34<27030>
71d46aa8dab5:6001:6001 [0] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.34<27030>
71d46aa8dab5:6002:6002 [1] NCCL INFO NET/Socket : Using [0]eth0:172.17.0.5<0>
71d46aa8dab5:6001:6001 [0] NCCL INFO NET/Socket : Using [0]eth0:172.17.0.5<0>
71d46aa8dab5:6001:6016 [0] NCCL INFO Setting affinity for GPU 0 to 0fffff

71d46aa8dab5:6001:6016 [0] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>

71d46aa8dab5:6001:6016 [0] bootstrap.cc:83 NCCL WARN NET/Socket : No usable listening interface found
71d46aa8dab5:6001:6016 [0] NCCL INFO bootstrap.cc:312 -> 2
71d46aa8dab5:6001:6016 [0] NCCL INFO init.cc:695 -> 2
71d46aa8dab5:6001:6016 [0] NCCL INFO init.cc:951 -> 2
71d46aa8dab5:6001:6016 [0] NCCL INFO misc/group.cc:69 -> 2 [Async thread]
71d46aa8dab5:6002:6015 [1] NCCL INFO Setting affinity for GPU 1 to 0fffff

71d46aa8dab5:6002:6015 [1] include/socket.h:197 NCCL WARN Net : No interface found in the same subnet as remote address 192.168.0.91<27030>

71d46aa8dab5:6002:6015 [1] bootstrap.cc:83 NCCL WARN NET/Socket : No usable listening interface found
71d46aa8dab5:6002:6015 [1] NCCL INFO bootstrap.cc:312 -> 2
71d46aa8dab5:6002:6015 [1] NCCL INFO init.cc:695 -> 2
71d46aa8dab5:6002:6015 [1] NCCL INFO init.cc:951 -> 2
71d46aa8dab5:6002:6015 [1] NCCL INFO misc/group.cc:69 -> 2 [Async thread]
Traceback (most recent call last):
  File "main.py", line 425, in <module>
    main()
  File "main.py", line 109, in main
    mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, args))
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 119, in join
    raise Exception(msg)
Exception: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap
    fn(i, *args)
  File "/home/gemfield/github/pytorch-examples/imagenet/main.py", line 151, in main_worker
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 285, in __init__
    self.broadcast_bucket_size)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 483, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1587428398394/work/torch/lib/c10d/ProcessGroupNCCL.cpp:514, unhandled system error, NCCL version 2.4.8
```

**Comment by gemfield** [comment]
@sjeaugey 

> I could only find this page : https://docs.docker.com/network/bridge/ which states : "in terms of Docker, a bridge network uses a software bridge which allows containers connected to the same bridge network to communicate".

Yes, this is exactly the docker bridge document, which ```bridge``` as the default network driver of Docker. But bridge can only work on same host, which is not  current situation.

Though I can use ```overlay``` ,```host``` or customized route table rules to have a woraround, again, I still want to make it work in default ```bridge``` mode. 

So,  the question somewhat becomes... is there a solution that we can customize the peer IPs that NCCL used?

**Comment by kwen2501** [comment]
So as the logs show and as @sjeaugey mentioned, the 192.168 address does not seem to be accessible from within the containers. That is, you don't have an interface that corresponds to that subnet. And I'd agree with @sjeaugey that this host address should be transparent to anything within the containers.

Regarding `NCCL_COMM_ID`, this environment would need to be the same across containers -- it is a rendezvous point.

**Comment by gemfield** [comment]
> So as the logs show and as @sjeaugey mentioned, the 192.168 address does not seem to be accessible from within the containers. 

This is not true, we can access 192.168 from container (i.e. CA/CB can access host A/ B);

> That is, you don't have an interface that corresponds to that subnet. 

Correct, this is exactly docker network isolation means.

I am thinking, why not force NCCL in container using ```192.168.0.91:27030``` since it is accessable from both containers?

**Comment by gemfield** [comment]
> Regarding NCCL_COMM_ID, this environment would need to be the same across containers -- it is a rendezvous point.

@kwen2501 I have fixed this and the similar errors still there.

**Comment by sjeaugey** [comment]
@gemfield the bootstrap is not the whole story. Even if you manage to get past that point, you'll still need containers to be able to talk to each other.

So, you need direct access between all your containers. As far as I understand, using bridge mode is just about giving a dedicated IP address to each container (instead of using the IP address of the host).

You can still route the traffic though. Granted that's a bit of work. That means setting up different subnets per node, then setting iptables to send traffic for a given subnet to a given node, then enabling ip forwarding on all nodes. 

But there is no other way. NCCL will not work if containers can't talk to each other. We do not have a network plugin that is capable of tunneling all the traffic through a port (which would be pretty terrible in terms of performance).

**Comment by gemfield** [comment]
@sjeaugey @kwen2501 .
Got it. Thanks very much.

**Comment by haruishi43** [comment]
@gemfield this is exactly the problem I'm facing right now. I have many nodes with exposed ports, but don't want to use `--network` since I want to use the nodes as ssh servers. Have you figured out how to use `torch.distributed` in the scenario you mentioned before?

**Comment by ShrimpLau** [comment]
FYI, after reading a little src code of NCCL communication setup progress and making some tests, each worker needs to bind a port in the node, using the IP ports available in system file `/proc/sys/net/ipv4/ip_local_port_range`(ref to [Troubleshooting](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html#ip-ports)), so there exists an unavoidable restriction that if we only bind one port between the container and the physical machine, the NCCL workers will bind ports which are invisible outside the container, therefore fail to establish a global connection, not to mention set up the topology.   

Meanwhile, there is another thing to notice at the same time, if we set up the `/proc/sys/net/ipv4/ip_local_port_range` too small, and we launch more workers than the available visible ports, there will be Warning: **nccl warn call to bind failed address already in use**. 

To summarize, the most direct way to use NCCL in the container is by setting the --network flag to host, so all the ports available will be transparent to containers.
